File: app/src/main/java/com/i2medier/financialpro/util/GraphUtils.java
package com.i2medier.financialpro.util;

import android.content.Context;
import android.graphics.Typeface;
import androidx.core.content.res.ResourcesCompat;

import com.i2medier.financialpro.R;
import com.github.mikephil.charting.charts.BarChart;
import com.github.mikephil.charting.charts.PieChart;
import com.github.mikephil.charting.components.AxisBase;
import com.github.mikephil.charting.components.Legend;
import com.github.mikephil.charting.components.XAxis;
import com.github.mikephil.charting.components.YAxis;
import com.github.mikephil.charting.data.BarData;
import com.github.mikephil.charting.data.BarDataSet;
import com.github.mikephil.charting.data.BarEntry;
import com.github.mikephil.charting.data.PieData;
import com.github.mikephil.charting.data.PieDataSet;
import com.github.mikephil.charting.data.PieEntry;
import com.github.mikephil.charting.formatter.PercentFormatter;
import com.github.mikephil.charting.formatter.ValueFormatter;
import com.github.mikephil.charting.interfaces.datasets.IBarDataSet;
import com.i2medier.financialpro.model.GraphModel;
import com.i2medier.financialpro.model.MonthModel;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;


public class GraphUtils {
    public static int COMON_GRAPH = 1;
    public static int MORTGAGE_GRAPH = 3;
    public static int SIMPLE_INTEREST_GRAPH = 2;
    ArrayList<String> PieEntryLabels;
    BarChart barChart;
    Context context;
    ArrayList<IBarDataSet> dataSets;
    ArrayList<PieEntry> entries;
    ArrayList<GraphModel> graphModelArrayList;
    int graphType;
    boolean isMonthWise;
    ArrayList<MonthModel> monthModels;
    PieChart pieChart;
    PieData pieData;
    PieDataSet pieDataSet;
    String MONTH_FORMATE = "MMM-yyyy";
    String YEAR_FORMATE = "yyyy";

    public GraphUtils(PieChart pieChart, ArrayList<GraphModel> arrayList, Context context) {
        this.pieChart = pieChart;
        this.graphModelArrayList = arrayList;
        this.context = context;
    }

    public GraphUtils(BarChart barChart, ArrayList<MonthModel> arrayList, Context context, boolean z, int i) {
        this.barChart = barChart;
        this.monthModels = arrayList;
        this.context = context;
        this.isMonthWise = z;
        this.graphType = i;
    }

    public void setupPieData() {
        Typeface font = ResourcesCompat.getFont(this.context, R.font.regular);
        this.entries = new ArrayList<>();
        this.PieEntryLabels = new ArrayList<>();
        this.pieDataSet = new PieDataSet(this.entries, "projects");
        this.pieChart.setUsePercentValues(true);
        this.pieChart.setDrawHoleEnabled(true);
        this.pieChart.setHoleColor(this.context.getResources().getColor(R.color.colorWhite));
        this.pieChart.setTransparentCircleAlpha(110);
        this.pieChart.setHoleRadius(58.0f);
        this.pieChart.setEntryLabelColor(this.context.getResources().getColor(R.color.colorDark));
        this.pieChart.setEntryLabelTypeface(font);
        Legend legend = this.pieChart.getLegend();
        legend.setVerticalAlignment(Legend.LegendVerticalAlignment.TOP);
        legend.setHorizontalAlignment(Legend.LegendHorizontalAlignment.RIGHT);
        legend.setOrientation(Legend.LegendOrientation.VERTICAL);
        legend.setDrawInside(false);
        legend.setEnabled(false);
        ArrayList arrayList = new ArrayList();
        for (int i = 0; i < this.graphModelArrayList.size(); i++) {
            this.entries.add(new PieEntry((float) this.graphModelArrayList.get(i).getValue(), this.graphModelArrayList.get(i).getLabel()));
            arrayList.add(Integer.valueOf(this.graphModelArrayList.get(i).getColor()));
        }
        this.pieDataSet.setColors(arrayList);
        this.pieDataSet.setValueLinePart1OffsetPercentage(80.0f);
        this.pieDataSet.setValueLinePart1Length(0.2f);
        this.pieDataSet.setValueLinePart2Length(0.4f);
        this.pieDataSet.setSliceSpace(3.0f);
        this.pieDataSet.setXValuePosition(PieDataSet.ValuePosition.OUTSIDE_SLICE);
        this.pieData = new PieData(this.pieDataSet);
        this.pieData.setValueFormatter(new PercentFormatter());
        this.pieData.setValueTextSize(12.0f);
        this.pieData.setValueTextColor(this.context.getResources().getColor(R.color.colorWhite));
        this.pieChart.setData(this.pieData);
        this.pieChart.getDescription().setEnabled(false);
        ((PieData) this.pieChart.getData()).setDrawValues(true);
        this.pieChart.animateY(3000);
        this.pieChart.invalidate();
    }

    public void setupPieData(String centerText) {
        setupPieData();
        if (centerText != null) {
            this.pieChart.setCenterText(centerText);
        }
    }

    public void setupBarChartData() {
        this.barChart.getAxisRight().removeAllLimitLines();
        this.barChart.getAxisLeft().removeAllLimitLines();
        this.barChart.clear();
        this.barChart.getAxisLeft().resetAxisMaximum();
        this.barChart.getAxisLeft().resetAxisMinimum();
        this.barChart.getAxisRight().resetAxisMaximum();
        this.barChart.getAxisRight().resetAxisMinimum();
        ArrayList arrayList = new ArrayList();
        ArrayList arrayList2 = new ArrayList();
        int i = this.graphType;
        if (i == COMON_GRAPH) {
            for (int i2 = 0; i2 < this.monthModels.size(); i2++) {
                float[] fArr = {(float) this.monthModels.get(i2).getPrincipalAmount(), (float) this.monthModels.get(i2).getInterest()};
                arrayList2.add(Integer.valueOf(this.context.getResources().getColor(R.color.graphcolor1)));
                arrayList2.add(Integer.valueOf(this.context.getResources().getColor(R.color.graphcolor2)));
                arrayList.add(new BarEntry(i2, fArr, arrayList2));
            }
        } else if (i == SIMPLE_INTEREST_GRAPH) {
            for (int i3 = 0; i3 < this.monthModels.size(); i3++) {
                float[] fArr2 = {(float) this.monthModels.get(i3).getTotalPrincipal(), (float) this.monthModels.get(i3).getTotalInterest()};
                arrayList2.add(Integer.valueOf(this.context.getResources().getColor(R.color.graphcolor1)));
                arrayList2.add(Integer.valueOf(this.context.getResources().getColor(R.color.graphcolor2)));
                arrayList.add(new BarEntry(i3, fArr2, arrayList2));
            }
        } else if (i == MORTGAGE_GRAPH) {
            for (int i4 = 0; i4 < this.monthModels.size(); i4++) {
                float[] fArr3 = {(float) this.monthModels.get(i4).getTotalPrincipal(), (float) this.monthModels.get(i4).getTotalInterest(), (float) this.monthModels.get(i4).getTotalTax()};
                arrayList2.add(Integer.valueOf(this.context.getResources().getColor(R.color.graphcolor1)));
                arrayList2.add(Integer.valueOf(this.context.getResources().getColor(R.color.graphcolor2)));
                arrayList2.add(Integer.valueOf(this.context.getResources().getColor(R.color.graphcolor3)));
                arrayList.add(new BarEntry(i4, fArr3, arrayList2));
            }
        }
        if (this.barChart.getData() != null && ((BarData) this.barChart.getData()).getDataSetCount() > 0) {
            ((BarDataSet) ((BarData) this.barChart.getData()).getDataSetByIndex(0)).setValues(arrayList);
            ((BarData) this.barChart.getData()).notifyDataChanged();
            this.barChart.notifyDataSetChanged();
            return;
        }
        BarDataSet barDataSet = new BarDataSet(arrayList, "Statistic");
        barDataSet.setDrawIcons(false);
        barDataSet.setColors(arrayList2);
        this.dataSets = new ArrayList<>();
        this.dataSets.add(barDataSet);
        BarData barData = new BarData(this.dataSets);
        barData.setBarWidth(0.5f);
        barData.setValueFormatter(new MyValueFormatter());
        barData.setValueTextColor(0);
        this.barChart.setData(barData);
        this.barChart.setScaleEnabled(false);
        this.barChart.getAxisLeft().setDrawGridLines(false);
        this.barChart.getXAxis().setDrawGridLines(false);
        this.barChart.getAxisLeft().setDrawLabels(true);
        this.barChart.setDrawBorders(false);
        this.barChart.setDrawGridBackground(false);
        XAxis xAxis = this.barChart.getXAxis();
        xAxis.setPosition(XAxis.XAxisPosition.BOTTOM);
        xAxis.setGranularityEnabled(true);
        xAxis.setEnabled(true);
        xAxis.setDrawGridLines(false);
        xAxis.setValueFormatter(new CustomXAxis(this.monthModels));
        xAxis.setDrawAxisLine(true);
        YAxis axisLeft = this.barChart.getAxisLeft();
        axisLeft.setEnabled(false);
        axisLeft.setDrawGridLines(false);
        axisLeft.setDrawAxisLine(false);
        YAxis axisRight = this.barChart.getAxisRight();
        axisRight.setEnabled(true);
        axisRight.setDrawGridLines(false);
        axisRight.setDrawAxisLine(false);
        axisLeft.setAxisMinimum(0.0f);
        axisRight.setAxisMinimum(0.0f);
        this.barChart.getDescription().setEnabled(false);
        this.barChart.getLegend().setEnabled(true);
        this.barChart.notifyDataSetChanged();
        this.barChart.setFitBars(true);
        this.barChart.getLegend().setEnabled(false);
        this.barChart.invalidate();
        ((BarData) this.barChart.getData()).notifyDataChanged();
        this.barChart.notifyDataSetChanged();
        this.barChart.setVisibleXRangeMaximum(10.0f);
        this.barChart.moveViewToX(this.monthModels.size());
    }

    public String getDate(Date date, String str) {
        return new SimpleDateFormat(str).format(date);
    }


    public class CustomXAxis extends ValueFormatter {
        List<MonthModel> monthModels;

        public CustomXAxis(List<MonthModel> list) {
            this.monthModels = list;
        }

        @Override
        public String getAxisLabel(float f, AxisBase axisBase) {
            List<MonthModel> list = this.monthModels;
            if (list != null) {
                int i = (int) f;
                if (list.get(i) == null) {
                    return "";
                }
                if (GraphUtils.this.isMonthWise) {
                    return GraphUtils.this.getDate(this.monthModels.get(i).getDateofGraph(), GraphUtils.this.MONTH_FORMATE);
                }
                return String.valueOf(this.monthModels.get(i).getYear());
            }
            return "";
        }
    }


    public class MyValueFormatter extends ValueFormatter {
        public MyValueFormatter() {
        }

        @Override
        public String getBarLabel(BarEntry barEntry) {
            float f = barEntry.getY();
            if (f <= com.github.mikephil.charting.utils.Utils.DOUBLE_EPSILON) {
                return "";
            }
            return "" + ((int) f) + " ";
        }
    }
}

File: app/src/main/java/com/i2medier/financialpro/util/GraphUtils.kt
package com.i2medier.financialpro.util

import android.content.Context
import androidx.core.content.ContextCompat
import androidx.core.content.res.ResourcesCompat
import com.github.mikephil.charting.charts.BarChart
import com.github.mikephil.charting.charts.PieChart
import com.github.mikephil.charting.components.AxisBase
import com.github.mikephil.charting.components.Legend
import com.github.mikephil.charting.components.XAxis
import com.github.mikephil.charting.components.YAxis
import com.github.mikephil.charting.data.BarData
import com.github.mikephil.charting.data.BarDataSet
import com.github.mikephil.charting.data.BarEntry
import com.github.mikephil.charting.data.PieData
import com.github.mikephil.charting.data.PieDataSet
import com.github.mikephil.charting.data.PieEntry
import com.github.mikephil.charting.formatter.PercentFormatter
import com.github.mikephil.charting.formatter.ValueFormatter
import com.github.mikephil.charting.interfaces.datasets.IBarDataSet
import com.i2medier.financialpro.R
import com.i2medier.financialpro.model.GraphModel
import com.i2medier.financialpro.model.MonthModel
import java.text.SimpleDateFormat
import java.util.ArrayList
import java.util.Date
import java.util.Locale

class GraphUtils {
    private var pieEntryLabels: ArrayList<String>? = null
    private var barChart: BarChart? = null
    private var context: Context
    private var dataSets: ArrayList<IBarDataSet>? = null
    private var entries: ArrayList<PieEntry>? = null
    private var graphModelArrayList: ArrayList<GraphModel>? = null
    private var graphType: Int = 0
    private var isMonthWise: Boolean = false
    private var monthModels: ArrayList<MonthModel>? = null
    private var pieChart: PieChart? = null
    private var pieData: PieData? = null
    private var pieDataSet: PieDataSet? = null
    private val MONTH_FORMATE = "MMM-yyyy"

    constructor(pieChart: PieChart, arrayList: ArrayList<GraphModel>, context: Context) {
        this.pieChart = pieChart
        this.graphModelArrayList = arrayList
        this.context = context
    }

    constructor(barChart: BarChart, arrayList: ArrayList<MonthModel>, context: Context, z: Boolean, i: Int) {
        this.barChart = barChart
        this.monthModels = arrayList
        this.context = context
        this.isMonthWise = z
        this.graphType = i
    }

    fun setupPieData() {
        val font = ResourcesCompat.getFont(context, R.font.regular)
        entries = ArrayList()
        pieEntryLabels = ArrayList()
        pieDataSet = PieDataSet(entries, "projects")

        val localPieChart = pieChart ?: return
        localPieChart.setUsePercentValues(true)
        localPieChart.setDrawHoleEnabled(true)
        localPieChart.setHoleColor(ContextCompat.getColor(context, R.color.colorWhite))
        localPieChart.setTransparentCircleAlpha(110)
        localPieChart.holeRadius = 58.0f
        localPieChart.setEntryLabelColor(ContextCompat.getColor(context, R.color.colorDark))
        localPieChart.setEntryLabelTypeface(font)

        val legend = localPieChart.legend
        legend.verticalAlignment = Legend.LegendVerticalAlignment.TOP
        legend.horizontalAlignment = Legend.LegendHorizontalAlignment.RIGHT
        legend.orientation = Legend.LegendOrientation.VERTICAL
        legend.setDrawInside(false)
        legend.isEnabled = false

        val arrayList = ArrayList<Int>()
        val graphList = graphModelArrayList.orEmpty()
        for (i in graphList.indices) {
            entries?.add(PieEntry(graphList[i].value.toFloat(), graphList[i].label))
            arrayList.add(graphList[i].color)
        }

        pieDataSet?.setColors(arrayList)
        pieDataSet?.valueLinePart1OffsetPercentage = 80.0f
        pieDataSet?.valueLinePart1Length = 0.2f
        pieDataSet?.valueLinePart2Length = 0.4f
        pieDataSet?.sliceSpace = 3.0f
        pieDataSet?.xValuePosition = PieDataSet.ValuePosition.OUTSIDE_SLICE

        pieData = PieData(pieDataSet)
        pieData?.setValueFormatter(PercentFormatter())
        pieData?.setValueTextSize(12.0f)
        pieData?.setValueTextColor(ContextCompat.getColor(context, R.color.colorWhite))

        localPieChart.data = pieData
        localPieChart.description.isEnabled = false
        (localPieChart.data as PieData).setDrawValues(true)
        localPieChart.animateY(3000)
        localPieChart.invalidate()
    }

    fun setupPieData(centerText: String?) {
        setupPieData()
        if (centerText != null) {
            pieChart?.centerText = centerText
        }
    }

    fun setupBarChartData() {
        val localBarChart = barChart ?: return
        val localMonthModels = monthModels ?: return

        localBarChart.axisRight.removeAllLimitLines()
        localBarChart.axisLeft.removeAllLimitLines()
        localBarChart.clear()
        localBarChart.axisLeft.resetAxisMaximum()
        localBarChart.axisLeft.resetAxisMinimum()
        localBarChart.axisRight.resetAxisMaximum()
        localBarChart.axisRight.resetAxisMinimum()

        val arrayList = ArrayList<BarEntry>()
        val arrayList2 = ArrayList<Int>()

        if (graphType == COMON_GRAPH) {
            for (i in localMonthModels.indices) {
                val fArr = floatArrayOf(
                    localMonthModels[i].PrincipalAmount.toFloat(),
                    localMonthModels[i].Interest.toFloat()
                )
                arrayList2.add(ContextCompat.getColor(context, R.color.graphcolor1))
                arrayList2.add(ContextCompat.getColor(context, R.color.graphcolor2))
                arrayList.add(BarEntry(i.toFloat(), fArr, arrayList2))
            }
        } else if (graphType == SIMPLE_INTEREST_GRAPH) {
            for (i in localMonthModels.indices) {
                val fArr = floatArrayOf(
                    localMonthModels[i].TotalPrincipal.toFloat(),
                    localMonthModels[i].TotalInterest.toFloat()
                )
                arrayList2.add(ContextCompat.getColor(context, R.color.graphcolor1))
                arrayList2.add(ContextCompat.getColor(context, R.color.graphcolor2))
                arrayList.add(BarEntry(i.toFloat(), fArr, arrayList2))
            }
        } else if (graphType == MORTGAGE_GRAPH) {
            for (i in localMonthModels.indices) {
                val fArr = floatArrayOf(
                    localMonthModels[i].TotalPrincipal.toFloat(),
                    localMonthModels[i].TotalInterest.toFloat(),
                    localMonthModels[i].TotalTax.toFloat()
                )
                arrayList2.add(ContextCompat.getColor(context, R.color.graphcolor1))
                arrayList2.add(ContextCompat.getColor(context, R.color.graphcolor2))
                arrayList2.add(ContextCompat.getColor(context, R.color.graphcolor3))
                arrayList.add(BarEntry(i.toFloat(), fArr, arrayList2))
            }
        }

        if (localBarChart.data != null && (localBarChart.data as BarData).dataSetCount > 0) {
            ((localBarChart.data as BarData).getDataSetByIndex(0) as BarDataSet).values = arrayList
            (localBarChart.data as BarData).notifyDataChanged()
            localBarChart.notifyDataSetChanged()
            return
        }

        val barDataSet = BarDataSet(arrayList, "Statistic")
        barDataSet.setDrawIcons(false)
        barDataSet.setColors(arrayList2)
        dataSets = ArrayList()
        dataSets?.add(barDataSet)

        val barData = BarData(dataSets)
        barData.barWidth = 0.5f
        barData.setValueFormatter(MyValueFormatter())
        barData.setValueTextColor(0)

        localBarChart.data = barData
        localBarChart.setScaleEnabled(false)
        localBarChart.axisLeft.setDrawGridLines(false)
        localBarChart.xAxis.setDrawGridLines(false)
        localBarChart.axisLeft.setDrawLabels(true)
        localBarChart.setDrawBorders(false)
        localBarChart.setDrawGridBackground(false)

        val xAxis = localBarChart.xAxis
        xAxis.position = XAxis.XAxisPosition.BOTTOM
        xAxis.isGranularityEnabled = true
        xAxis.isEnabled = true
        xAxis.setDrawGridLines(false)
        xAxis.valueFormatter = CustomXAxis(localMonthModels)
        xAxis.setDrawAxisLine(true)

        val axisLeft: YAxis = localBarChart.axisLeft
        axisLeft.isEnabled = false
        axisLeft.setDrawGridLines(false)
        axisLeft.setDrawAxisLine(false)

        val axisRight: YAxis = localBarChart.axisRight
        axisRight.isEnabled = true
        axisRight.setDrawGridLines(false)
        axisRight.setDrawAxisLine(false)

        axisLeft.axisMinimum = 0.0f
        axisRight.axisMinimum = 0.0f

        localBarChart.description.isEnabled = false
        localBarChart.legend.isEnabled = true
        localBarChart.notifyDataSetChanged()
        localBarChart.setFitBars(true)
        localBarChart.legend.isEnabled = false
        localBarChart.invalidate()
        (localBarChart.data as BarData).notifyDataChanged()
        localBarChart.notifyDataSetChanged()
        localBarChart.setVisibleXRangeMaximum(10.0f)
        localBarChart.moveViewToX(localMonthModels.size.toFloat())
    }

    fun getDate(date: Date, str: String): String {
        return SimpleDateFormat(str, Locale.getDefault()).format(date)
    }

    inner class CustomXAxis(private val monthModels: List<MonthModel>) : ValueFormatter() {
        override fun getAxisLabel(f: Float, axisBase: AxisBase?): String {
            val i = f.toInt()
            if (i < 0 || i >= monthModels.size) {
                return ""
            }
            return if (isMonthWise) {
                getDate(monthModels[i].getDateofGraph(), MONTH_FORMATE)
            } else {
                monthModels[i].year.toString()
            }
        }
    }

    inner class MyValueFormatter : ValueFormatter() {
        override fun getBarLabel(barEntry: BarEntry): String {
            val f = barEntry.y
            return if (f <= com.github.mikephil.charting.utils.Utils.DOUBLE_EPSILON) {
                ""
            } else {
                "${f.toInt()} "
            }
        }
    }

    companion object {
        @JvmField
        var COMON_GRAPH: Int = 1

        @JvmField
        var SIMPLE_INTEREST_GRAPH: Int = 2

        @JvmField
        var MORTGAGE_GRAPH: Int = 3
    }
}
